### DISCUSSION TITLE
Adding scripts to the <head> without head.js

### DISCUSSION THREAD

**Discussion by jack-pallot** [vercel/next.js#47166] [Help]
### Summary

Now that head.js has been deprecated in 13.2, the new method to add items to the `<head>` section of your app is with the metadata API. Unfortunately, I'm trying to get an analytics script to appear before the closing head tag (a requirement), but I have been completely unsuccessful in doing so. Some things I have tried include:

- Using next/script and a normal script tag
- Add a script to layout.js
- Adding a script to page.js and trying both beforeInteractive and afterInteractive
- Creating a server-only component with the script tag in
- Manually adding `<head>` tags around the script
- Manually adding `<head>` tags around the server component

None of these work, which begs the question... Why deprecate page.js when this seems like the only way to do something like this? Are there any known workarounds?

**Comment by icyJoseph** [comment]
Interesting requirement, I don't have a solution, but I am curious, why should it go last? Is it some kind of loading ordering that's expected to happen? or should be the last thing in the deferred script queue?

**Comment by jack-pallot** [comment]
It just needs to go before the closing `<head>` tag. This is for a Google Tag Manager script and it specifies it needs to be installed as high up as possible, or you risk missing some tracked events. I know this isn't recommended practice in reality, but frustratingly the auditing tools used in our marketing team seem to raise it as an issue. This was definitely possible before, but doesn't seem to be now.

**Comment by hexionaut** [comment]
Can confirm this is problematic for Google Tag Manager ...

**Comment by icyJoseph** [comment]
I was able to do this at the root layout:

```tsx
import Script from "next/script";
import "./globals.css";

export const metadata = {
  title: "Create Next App",
  description: "Generated by create next app",
};

export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <html lang="en">
      <head>
        <Script
          src="https://cdnjs.cloudflare.com/ajax/libs/lodash.js/4.17.21/lodash.min.js"
          strategy="beforeInteractive"
        />
      </head>
      <body>{children}</body>
    </html>
  );
}
```

Which rendered:

![Screenshot 2023-03-21 at 13 08 59](https://user-images.githubusercontent.com/21013447/226601650-cbbe541b-c6f3-43a8-a0d4-432293618781.png)

I went about it from the docs:

> Good to know: Scripts with beforeInteractive will always be injected inside the head of the HTML document regardless of where it's placed in the component.

Though the examples in the beta docs do not really work:

- https://beta.nextjs.org/docs/api-reference/components/script#beforeinteractive

That one kept giving me errors.

**Comment by hexionaut** [comment]
Thank you @icyJoseph ! This is also what I tried at first (the beta doc one) but had errors too.

Didn't know `head` content get merged. Now I know. 

Last point, TypeScript is dropping some warnings : 

```
Do not use `<head>` element. Use `<Head />` from `next/head` instead. See: https://nextjs.org/docs/messages/no-head-element

`next/script`'s `beforeInteractive` strategy should not be used outside of `pages/_document.js`. See: https://nextjs.org/docs/messages/no-before-interactive-script-outside-document
```

**Reply by icyJoseph** [reply]
Those are lint errors, not typescript. I updated my eslint-next-config to latest, and only see the _document warning, which is something you can ignore for now. This is what you get into when you choose to use technologies in beta stage 😅  hopefully this isn't some production stuff for you.

**Reply by hexionaut** [reply]
You are right. My bad ! Thanks for helping m8 !

**Comment by jack-pallot** [comment]
Thanks for your input and giving it a test. I think the issue is with inline scripts. I can't get the following to ever be in the `<head>` section.

`<Script id="script-gtm" strategy="beforeInteractive">
  {'console.log('tester')'}
  </Script>`

**Reply by icyJoseph** [reply]
Aha, what about using `dangerouslySetInnerHTML`?

```jsx
<Script
  id="show-banner"
  strategy="afterInteractive"
  dangerouslySetInnerHTML={{
    __html: `document.getElementById('banner').classList.remove('hidden')`,
  }}
/>
```

**Reply by jack-pallot** [reply]
Yeah that works, thanks for the help! One thing though... I get a hydration error is if the script is inside of a layout.js file. I'm not sure if that's intentional or not?

**Reply by icyJoseph** [reply]
Aha, I don't get any hydration errors. Could you share some more code? What does the stack say? Maybe you need to update Next.js? 

Also please don't forget to mark my other answer and answered (if you think it did).

**Reply by byron-ino** [reply]
hey @jack-pallot did you ever find the cause of your hydration errors? Currently seeing similar issues when simply adding a script to the root layout.js

**Reply by ramazansancar** [reply]
Solved my problem for Partnero integration. Thanks. @icyJoseph

**Comment by housseindjirdeh** [comment]
`beforeInteractive` scripts should be injected into `<head>`, but I'm also seeing hydration errors and/or the script failing to execute when including to my root layout.

We'll submit a fix for this soon. #51242 to track it!

**Comment by dayanrr91** [comment]
I did it like this directly and it worked

![image](https://github.com/user-attachments/assets/d31bdcd0-2e30-45ad-ac80-c9b6ca212a69)

**Comment by jeromeweiss** [comment]
Based on the Next.js documentation (https://nextjs.org/docs/pages/building-your-application/optimizing/scripts) you can use the `<Script />` tag to add scripts (e.g. for google analytics).

It's important to add the property `strategy='beforeInteractive'` to the script when it should be injected into the header and loaded before any Next.js module (https://nextjs.org/docs/pages/api-reference/components/script#strategy).

**Comment by S-Nishant** [comment]
strategy='beforeInteractive'  attribute on next/Script adds your script it to <head> tag

**Comment by elepner** [comment]
We've got a requirement from Google Analytics team that consent command must be placed right below opening ``<head>`` tag. But nextjs inserts its own script tags regardless. I understand that beforeInteractive would be executed before any nextjs code but Google analytics team wants this code to be in the specific place.
